README.txt
rbridge
ruby-mnesia.rubyforge.com
by Toshiyuki Hirooka, Chuck Vose

== DESCRIPTION:

RBridge allows the use of Erlang code within a Ruby program. Optionally allowing
asynchronous access and ruby-like syntax or by using Erlang code directly when
more expressive or complex code is needed.

== FEATURES/PROBLEMS:

* Allows use of Erlang code within Ruby
* Doesn't yet allow Ruby code within Erlang

== SYNOPSIS:

  1. Start an instance of the rulang server by using bin/rulang

  2. Execute your ruby code by connecting (as shown below) to either the RBridge class if you want
  asynchronous access and ruby style code access, or by connecting to the Erlang
  class if you need more complicated code evaluation.

  # load the library
  require 'rbridge'

  # Connect to localhost on port 9900 and using only the erlang module。
  r_erlang = RBridge.new("erlang", "localhost", 9900)

  # This is equivalent to erlang:length([1,2,3,4,5,6,7,8,9]).
  p r_erlang.length([1,2,3,4,5,6,7,8,9])
  # => 9

  # Run Erlang code directly
  a = RBridge.new(nil, "localhost", 9900)
  p a.erl("10*10.")
  # => 100

== REQUIREMENTS:

* Erlang (erlang.org) -
  There are current windows binaries as well as source for other oses. Installing
  on OS X should just need: ./configure --prefix=/opt/local && make && sudo make install.

* Ruby (ruby-lang.org) -
  On the downloads page if you skip past the source downloads there are a lot
  of one-click installers for various oses. If you go with source you at least
  need to install rubygems from rubygems.org.

== INSTALL:

* sudo gem install rbridge

== LICENSE:

RBridge - A Ruby to Erlang bridge allowing the use of Erlang commands within 
normal Ruby code.
Copyright (C) 2008 Toshiyuki Hirooka <toshi.hirooka@gmail.com>, Chuck Vose <vosechu@gmail.com>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


History.txt
== 0.2.0 / 2008-04-24 (vosechu)
* 1 major enhancement
  * Actually runs headless via the rulang command. 
  
* 3 minor enhancements
  * Uses optparse for clearer command-line options.
  * Option to specify port, compile directory, mnesia directory, sname, and location of server file.
  * Some error checking, and debug output added. Checks to see if server is already running on specified port.

== 0.1.2 / 2008-01-10 (vosechu)
* major enhancements
  * Runs headless by using the rulang command or via the command-line using the
    -noshell switch to erl. 
  
* bugs / limitations
  * rulang command isn't ready for prime time at all

== 0.1.1 / 2008-01-07 (vosechu)
* 2 minor enhancements
  * auto-detect whether someone has put a period at the end of a set of exprs
  * updated specs catch a few more edge cases

== 0.1 / 2008-01-04 (vosechu)

* 1 major enhancement
  * Allows many-lined Erlang commands when using the Erlang class for direct
    access. 
* 3 minor enhancements
  * Revised error checking to allow custom resolution
  * stop_server/0 now shuts down any servers running in an erl shell
  * created tests to run with RSpec and autotest (erl server must be running 
    on port 9900).
* 3 other changes
  * Forked onto rubyforge complete with rdocs, gem, etc.
  * Translated some comments into english as well as the README
  * Packaged explicitly with gpl documentation

spec/rulang_spec.rb
# RBridge - A Ruby to Erlang bridge allowing the use of Erlang commands within 
# normal Ruby code.
# Copyright (C) 2008 tosik <toshi.hirooka@gmail.com>, Chuck Vose <vosechu@gmail.com>
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# require 'rubygems'
# require 'rbridge'
require 'spec/spec_helper'

describe "RBridge (erlang module)" do
  before(:each) do
    @r_erlang = RBridge.new("erlang", "localhost", 9900)
  end
  
  it "should compute the length of an 8 slot array" do
    @r_erlang.length([1,2,3,4,5,6,7,8,9]).should eql(9)
  end
  
  it "should compute the length of a 14 character string" do
    @r_erlang.length("rulang bridge!").should eql(14)
  end
end

describe "RBridge (erl_syntax module)" do
  before(:each) do
    @r_erl_syntax = RBridge.new("erl_syntax", "localhost", 9900)
  end
  
  it "should return the syntax tree of the term \"test\"" do
    @r_erl_syntax.abstract("test").should eql(["tree", "string", ["attr", 0, [], "none"], "test"])
  end
end

describe "RBridge (asynchronous erlang module)" do
  before(:each) do
    @ar_erlang = RBridge.new("erlang", "localhost", 9900, true)
  end
  
  it "should display three messages in the order they're executed" do
    out = ""
    
    thread = @ar_erlang.length([6,3,5,2,4,1], proc { |x|
      out << "#{x}"
      out << "second"
    })
    out << "first"
    thread.join
    
    out.should eql("first6second")
  end
  
  it "should display three messages in the order they're executed" do
    out = ""
    
    thread = @ar_erlang.erl("erlang:length([6,3,5,2,4,1]).", proc { |x|
      out << "#{x}"
      out << "second"
    })
    out << "first"
    thread.join
    
    out.should eql("first6second")
  end
end

describe "Direct Erlang code" do
  before(:each) do
    @e_raw = RBridge.new(nil, "localhost", 9900)
  end
  
  it "should evaluate mathematical expressions" do
    @e_raw.erl("10*10.").should eql(100)
  end
  
  it "should evaluate multiple expressions seperated by commas" do
    @e_raw.erl("QH = [1,2,3,4,5], erlang:length(QH).").should eql(5)
  end
  
  it "should add a period if there isn't one at the end of an expression" do
    @e_raw.erl("QH = [1,2,3,4,5], erlang:length(QH)").should eql(5)
  end
end

spec/spec_helper.rb
# Copyright (C) 2008 tosik <toshi.hirooka@gmail.com>, Chuck Vose <vosechu@gmail.com>
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

require 'lib/vosetools'
include Vosetools

exclude = ['spec/spec_helper.rb', 'lib/rulang.erl', 'lib/rulang.beam']

Dir.glob('lib/*').each do |file|
  unless exclude.include?(file)
    require file
  end
end

# Dir.glob('spec/*').each do |file|
#   unless exclude.include?(file)
#     require file
#     # log_p "requiring #{file}"
#   end
# end

lib/vosetools.rb
# # These files may not be used in this project but are merely notes collected
# # and inserted into all of Chuck Vose's projects. Please disregard them 
# # unless they help you.

# # PLUGINS:
# # Autotest with UnitRecord
# # By default, running autotest in the Rails directory will run the unit tests. 
# # To run the functional tests, do: AUTOTEST='functional' autotest
# ruby script/plugin install http://svn.soniciq.com/public/rails/plugins/iq_autotest
# # Query Analyzer to see what we're doing under the surface
# # Courtesy of Bob Silva at http://agilewebdevelopment.com/plugins/query_analyzer
# ruby script/plugin install http://svn.nfectio.us/plugins/query_analyzer
# # Dev mode performance cacher by Josh Goebel
# # Caches dev code after file changes to make refreshes happen in the background
# ruby script/plugin install http://svn.techno-weenie.net/projects/plugins/dev_mode_performance_fixes/
# # Rspec for running BDD
# ruby script/plugin install svn://rubyforge.org/var/svn/rspec/tags/CURRENT/rspec
# ruby script/plugin install svn://rubyforge.org/var/svn/rspec/tags/CURRENT/rspec_on_rails

# # GEMS:
# # UnitRecord disconnects tests from the database and restructures tests into
# # Proper unit/functional tests. (Unit == no database, no externals). Run 
# # rake vose:convert_to_unitrecord[_and_commit] after installing.
# sudo gem install unit_record
# # Zentest provides test generation and autotest for hands free testing
# # RSpec ~/.autotest file from http://reinh.com/2007/9/6/rspec-autotest-and-growl-oh-my
# # Test::Unit ~/.autotest file from 
# sudo gem install zentest
# # Dust provides some handy helper methods to testing
# sudo gem install dust
# # Mocha allows cross unit tests by mocking and stubbing functions
# sudo gem install mocha

require 'rubygems'
# require 'mocha'
# require 'stubba'
# require 'dust'

# require 'factory'

module Vosetools
  def log_p(message)
    require 'pp'
    
    p "*********************"
    pp message
    p "*********************"
    
    # if !message.is_a?(Array)
    #   `growlnotify -m "#{message}"`
    # end
  end
  
  def profile
    require 'ruby-prof'
    
    result = RubyProf.profile do
      yield
    end
    
    # Print a flat profile to text
    printer = RubyProf::FlatPrinter.new(result)
    printer.print(STDOUT, 5)
  end
  
  # def has_a(obj)
  #   class_eval do
  #     define_method(:method_missing) do |*values|
  #       if obj.methods.include?(values[0])
  #         log_p values
  #         obj.send!(:values[0], values[1..(values.length-1)])
  #       end
  #     end
  #   end
  # end
end

# # Adapted from http://sami.samhuri.net/2007/4/11/activerecord-base-find_or_create-and-find_or_initialize
# # Originally created by sjs
# module ActiveRecordExtensions
#   def self.included(base)
#     base.extend(ClassMethods)
#   end
#   
#   module ClassMethods
#     def find_or_create(params)
#       begin
#         return self.find(params[:id])
#       rescue ActiveRecord::RecordNotFound => e
#         attrs = {}
#         
#         # search for valid attributes in params
#         self.column_names.map(&:to_sym).each do |attrib|
#           # skip unknown columns, and the id field
#           next if params[attrib].nil? || attrib == :id
# 
#           attrs[attrib] = params[attrib]
#         end
# 
#         # call the appropriate ActiveRecord finder method
#         found = self.send("find_by_#{attrs.keys.join('_and_')}", *attrs.values) if !attrs.empty?
#         
#         if found && !found.nil?
#           return found
#         else
#           return self.create(params)
#         end
#       end
#     end
#     alias create_or_find find_or_create
#   end
# end
# 
# ActiveRecord::Base.send(:include, ActiveRecordExtensions)

# Copyright: Jay Philips
# jicksta.com/articles/2007/08/04/the-methodphitamine
module Kernel
  protected
  def it() It.new end
  alias its it
end

class It

  undef_method(*(instance_methods - %w*__id__ __send__*))

  def initialize
    @methods = []
  end

  def method_missing(*args, &block)
    @methods << [args, block] unless args == [:respond_to?, :to_proc]
    self
  end

  def to_proc
    lambda do |obj|
      @methods.inject(obj) do |current,(args,block)|
        current.send(*args, &block)
      end
    end
  end
end

lib/erlang_adapter.rb
# RBridge - A Ruby to Erlang bridge allowing the use of Erlang commands within 
# normal Ruby code.
# Copyright (C) 2008 Toshiyuki Hirooka <toshi.hirooka@gmail.com>, Chuck Vose <vosechu@gmail.com>
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#
# Erlang specific commands and evaluator
#
class ErlangAdapter

	# [Host]
  # Erlang host server name (the default is "localhost")
  # [Port]
  # Erlang server port number (the default is "9900")
	def initialize(host="localhost", port=9900)
		@host, @port = host, port
	end

	# Send commands to the Erlang server to be evaluated. Must have trailing 
  # whitespace to differentiate between floats and EOL.
	def eval(command)		
		socket = TCPSocket.new(@host, @port)
				
    # There has to be a trailing space at the end of a command for erl_scan:tokens
    # to process the string.
    command << '.' if command.scan(/\.\w*/).empty?
		command << " "
		socket.write(command)
		socket.read # ...?
	end

	# Create an Erlang command from the ruby-style syntax
	# [erlang_mod]
	#   Erlang module
	# [erlang_func]
	#   Erlang function
	# [args]
	#   Arguments to pass to function. Should be in a Ruby array. 
	def make_command(erlang_mod, erlang_func, args)
		"#{erlang_mod}:#{erlang_func}(#{to_erlang_args(args)})."
	end

	# Translate a Ruby array to Erlang elements.
	def to_erlang_args(args)
		args.map{ |x| to_erlang_literal x }.join(',')
	end

  # TODO: Make this handle hashes
	# Convert Ruby objects to Erlang objects. Throw an error if the object passed
  # isn't a basic argument. 
	def to_erlang_literal(x)
		if x.kind_of?(String) then
			"\"#{x}\""
		elsif x.kind_of?(Integer) then
			x
		elsif x.kind_of?(Float) then
			x
		elsif x.kind_of?(Array) then
			"[#{x.map{|y| to_erlang_literal(y)}.join(',')}]"
		else
			raise "Can't use \"#{x.type} Object\" in arguments."
		end
	end

	# Quick boolean to decide whether the data returned from Erlang was an error.
	def is_error(str)
		str[0..5] == "Error:"
	end
end


rulang_server/rulang.erl
%% -*- erlang -*-
% RBridge - A Ruby to Erlang bridge allowing the use of Erlang commands within 
% normal Ruby code.
% Copyright (C) 2008 Toshiyuki Hirooka <toshi.hirooka@gmail.com>, Chuck Vose <vosechu@gmail.com>
% 
% This program is free software; you can redistribute it and/or
% modify it under the terms of the GNU General Public License
% as published by the Free Software Foundation; either version 2
% of the License, or (at your option) any later version.
% 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
-module(rulang).
-export([start_server/1, stop_server/0, loop/1, handle_connection/1, main/1, usage/0]).
-include_lib("stdlib/include/qlc.hrl").

main([Port]) ->
  try
    mnesia:create_schema([node()]),
    mnesia:start(),
    start_server(Port)
  catch
    _:_ ->
    % TODO: This should report the error as well as the usage
      usage()
  end;
main(_) ->
  usage().
  
usage() ->
  io:format("./rulang [port number]").

% Start a server and connect to the port specified by the bound variable Port.
% If all goes according to plan start the rulang loop and display a happy message.
start_server(Port) ->
	case gen_tcp:listen(Port, [binary, {packet, 0}, {active, true}]) of
	  {ok, ListenSocket} ->
    	io:format("[RBridge Server Started]~n"),
    	loop(ListenSocket);
    {error, Reason} ->
      io:format(Reason)
  end.

% Just kill the parent thread.
stop_server() ->
  io:format("[RBridge Server Stopped]~n"),
	exit(self()).

% Loop forever waiting for connections to wake us and either spawn a thread to
% deal with commands or die if the connection is broken.
loop(ListenSocket) ->
  % Block until we have a connection waiting
	case gen_tcp:accept(ListenSocket) of
	  
    % If the connection is valid read the commands and attempt to eval them.
    {ok, Socket} -> 
      spawn(fun() ->
        handle_connection(Socket)
      end),
    
    % Continue the loop
  	loop(ListenSocket)
  end.
	
% Wrapper that catches errors and closes the socket after we're done with it.
handle_connection(Socket) ->
  % Try to execute the commands in the tcp socket.
  try communication(Socket)
  % But catch any errors and pass them along to the tcp socket so we can see 
  % them in ruby.
  catch
    error:Reason ->
      {gen_tcp:send(Socket, io_lib:format("Error: ~p~n", [Reason]))}
  end,
  % Finally make sure the socket gets closed.
  ok = gen_tcp:close(Socket).

% Try to evaluate the code submitted throwing an exception if the evaluation
% doesn't work. 
communication(Socket) ->
  % Read the commands from the tcp socket
	{ok, Binary} = gen_tcp:recv(Socket, 0),
  % Run the commands through the erlang eval function
	{ok, Result} = eval(binary_to_list(Binary)),
  % Send the results back through the tcp pipe attempting to reconstitute the
  % Erlang values back into Ruby data structures along the way. 
	gen_tcp:send(Socket, io_lib:format("~p~n", [to_ruby(Result)])).
	
% Parse the commands into Erlang terms and run them. Return the result of the
% evaluation.
eval(Expression) ->
	{done, {ok, Scanned, _}, _} = erl_scan:tokens([], Expression, 0),
	{ok, Parsed} = erl_parse:parse_exprs(Scanned),
	{value, Result, _} = erl_eval:exprs(Parsed, []),
	{ok, Result}.

% Change Erlang values into Ruby values.
to_ruby(Expression) when is_list(Expression) ->
	lists:map(fun(X) -> to_ruby(X) end, Expression);
	%list_to_binary("'" ++ Expression ++ "'");
to_ruby(Expression) when is_tuple(Expression) ->
	to_ruby(tuple_to_list(Expression));
to_ruby(Expression) when is_binary(Expression) ->
	to_ruby(binary_to_list(Expression));
to_ruby(Expression) when is_integer(Expression) ->
	Expression;
to_ruby(Expression) when is_float(Expression) ->
	Expression;
to_ruby(Expression) when is_atom(Expression) ->
	atom_to_list(Expression).

lib/rbridge.rb
# RBridge - A Ruby to Erlang bridge allowing the use of Erlang commands within 
# normal Ruby code.
# Copyright (C) 2008 Toshiyuki Hirooka <toshi.hirooka@gmail.com>, Chuck Vose <vosechu@gmail.com>
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

require "socket"
require "erlang_adapter"

#
# Erlang サーバの関数を Ruby のメソッドのように扱えるようにするクラス
#
class RBridge
  VERSION = "0.2.0"
  
  # [Mod]
  # Erlang module to use when using ruby-style access (the default is "erlang")
	# [Host]
  # Erlang host server name (the default is "localhost")
  # [Port]
  # Erlang server port number (the default is "9900")
  # [Async]
  # Whether asynchronous communication (defaults to "false")
	def initialize(mod, host="localhost", port=9900, async=false)
    # Which module should we call by default? If nothing is specified try the
    # erlang module.
		@mod = mod
    @mod = "erlang" if @mod.nil?
    
		@erlang = ErlangAdapter.new(host, port)
		@async = async
	end

	# Ruby-style method calling. 
	def method_missing(symbol, *args)

		# Pop off the block used for asynchronous access.
		block = args.pop if @async == true

		# Try to create a command that Erlang knows what to do with from the ruby
    # style syntax we're given.
		command = @erlang.make_command(@mod, symbol, args)
		
		erl(command, block)
	end
	
  # def sql_erl(command, block=nil)
  #   res = command.scan(/[^\ ]+/)
  #   case res[0]
  #     when /select/i
  #       erl_command = "Q = qlc:q([X || X <- mnesia:table(#{res[3]})]),
  #                      F = fun() -> qlc:e(Q) end, 
  #                      {atomic, Val} = mnesia:transaction(F), 
  #                      Val."
  #     else
  #       erl_command = nil
  #     end
  #   
  #   puts erl_command
  #   erl(erl_command, block)
  #   end
	
  # Send commands to Erlang to be processed.
  def erl(command, block=nil)
    # log_p command   
    begin
      if @async == true then
        # If we're asynchronous generate a thread around the call then pass 
          # the results back to a block to display.
        Thread.new do
          res = @erlang.eval(command)
          raise res if @erlang.is_error(res)
          block.call(eval(res))
        end
      elsif @async == false then
        # Blocking mode should just eval and wait for data to come back.
        res = @erlang.eval(command)
        raise res if @erlang.is_error(res)
        eval res
      end
    rescue => res
      raise "[Error]=>#{res}"
    end
  end
  
  # def tv
  #   erl("tv:start().")
  # end
end
